### ISSUE TITLE
Folding output (at least in GHA/travis)

### ISSUE THREAD

**Issue by henryiii** [joerick/cibuildwheel#349]
Feature request:

The current output shows up as one long section, but recently GitHub Actions have started showing up with folding support, which is really quite nice. (See https://github.com/peaceiris/actions-gh-pages/blob/master/CHANGELOG.md) I'm sure I first saw it in the docs, where they list both the echo method and a javascript method for custom Actions, but can't find it now; here's a page that comes up from a google search: https://github.community/t5/GitHub-Actions/Has-github-action-somthing-like-travis-fold/td-p/37715. Travis also has a fold feature, and other systems might have them as well.

I would like to see the logs folded per wheel, maybe wheel and test separated. Just thinking about a possible design idea: there could be a set of variables, such as FOLD_START and FOLD_END, that would be printed (with templating, so at the very least the name of the current arch being built is available. Something like this would then be the setup for GitHub Actions:

```bash
CIBW_FOLD_START: "::group::{action} {wheel}"
CIBW_FOLD_END: "::groupend::"
```

Then the logs would look like:

```
> Setup all
> Building cp37-cp37m-macosx_10_9_x84_64
> Testing cp37-cp37m-macosx_10_9_x84_64
```

Maybe the defaults for GitHub Actions could be applied if it detected GHA (don't know how much customization you want per service), or this should just be something you set in examples.

What do you think? I can probably help with the implementation if you think it is a good idea.

PS: This may be called subfold, since each action folds.

**Comment by joerick** [comment]
Love this idea! I've been thinking for a while that it would be lovely to clean up our logs and folding might be just be the perfect way to do it.

My thinking on logs (before considering folding) was something like this: 

![Artboard](https://user-images.githubusercontent.com/1244307/82752720-f5678680-9db7-11ea-9ab6-0710ca225ee4.png)

But the obvious criticism of the above is 'what if I want to see the output of my before_build...', so maybe we could put that output into folding groups instead of hiding entirely.

In terms of how we'd get to the above, I can see something like the above be quite tricky to pull off with our current Docker build in bash. So something that fancy might have to wait until we have more flow control in the host Python (`docker exec` commands into a running container).

So maybe an interim solution would be good, which just starts a fold group at the start of a build - we already have this level of flow control in Python today. That would look something like:

![Artboard Copy 2](https://user-images.githubusercontent.com/1244307/82753048-8e979c80-9dba-11ea-8711-d492dd99e923.png)

Then the user would have to click into a group to see more on how it failed. This feels like something feasible that we could implement right now.

In terms of how this might manifest in terms of options, perhaps something like `CIBW_VERBOSE`/`CIBW_OUTPUT_FORMAT`, with options like-

- `quiet` Don't print any build output, except on failure
- `fold` (default) Fold all build output, if supported on the build platform (as detected by environment variables). If folding not supported, print build output as normal.
- `all` Print all build output, always, and don't fold.

---

btw: the syntax on travis can be seen [here](https://travis-ci.org/github/macbre/travis-fold). click 'raw log' to see the tags. Looks like each fold starts with `travis_fold:start:IDENTIFIER`, on the line before the 'header' of the group.

**Comment by henryiii** [comment]
Just a tiny bit of context; I'm using PEP 517 builds with the first verbose level, and when that on PIP prints a ***ton*** of information - one line for every package it considers (that's a few thousand lines just for checking every version of setuptools ever released, for example), making the logs *very* hard to look over for compiler warnings. Having folding would really help with that (even if it was just something easy to search for by hand would still be helpful). Turning off verbosity / hiding build output would hide the compiler warnings too.

(I've used travis folding before, probably for Plumbum or CLI11, which is why I know it does support it. It wasn't as reliable as GHA folding, at least back then, but it worked.)

**Comment by YannickJadoul** [comment]
@joerick managed to do this in #458, and released it in 1.7.0! :-)
